src/algebra/order/pointwise.lean

```diff
@@ -1,20 +1,141 @@
+by { rw ←image_inv, exact ((order_iso.inv α).map_cSup' hs₀ hs₁).symm }
+
+end conditionally_complete_lattice
+
+variables [complete_lattice α] [covariant_class α α (*) (≤)] [covariant_class α α (swap (*)) (≤)]
+
+@[to_additive] lemma Sup_inv (s : set α) : Sup s⁻¹ = (Inf s)⁻¹ :=
+by { rw [←image_inv, Sup_image], exact ((order_iso.inv α).map_Inf _).symm }
+
+@[to_additive] lemma Inf_inv (s : set α) : Inf s⁻¹ = (Sup s)⁻¹ :=
+by { rw [←image_inv, Inf_image], exact ((order_iso.inv α).map_Sup _).symm }
+
+end group
+
+section comm_group
+variables [comm_group α]
+
+section conditionally_complete_linear_order
+variables [conditionally_complete_linear_order α] [covariant_class α α (*) (≤)] {s t : set α}
+
+@[to_additive] lemma cSup_mul (hs₀ : s.nonempty) (hs₁ : bdd_above s) (ht₀ : t.nonempty)
+  (ht₁ : bdd_above t) :
+  Sup (s * t) = Sup s * Sup t :=
+begin
+  refine cSup_eq_of_forall_le_of_forall_lt_exists_gt (hs₀.mul ht₀) _ (λ a ha, _),
+  { rintro f ⟨a, b, ha, hb, rfl⟩,
+    exact mul_le_mul' (le_cSup hs₁ ha) (le_cSup ht₁ hb) },
+  { obtain ⟨b, hb, hab⟩ := exists_lt_of_lt_cSup hs₀ (div_lt_iff_lt_mul.2 ha),
+    obtain ⟨c, hc, hbc⟩ := exists_lt_of_lt_cSup ht₀ (div_lt''.1 hab),
+    exact ⟨b * c, mul_mem_mul hb hc, div_lt_iff_lt_mul'.1 hbc⟩ }
+end
+
+@[to_additive] lemma cInf_mul (hs₀ : s.nonempty) (hs₁ : bdd_below s) (ht₀ : t.nonempty)
+  (ht₁ : bdd_below t) :
+  Inf (s * t) = Inf s * Inf t :=
+begin
+  refine cInf_eq_of_forall_ge_of_forall_gt_exists_lt (hs₀.mul ht₀) _ (λ a ha, _),
+  { rintro f ⟨a, b, ha, hb, rfl⟩,
+    exact mul_le_mul' (cInf_le hs₁ ha) (cInf_le ht₁ hb) },
+  { obtain ⟨b, hb, hab⟩ := exists_lt_of_cInf_lt hs₀ (lt_div_iff_mul_lt.2 ha),
+    obtain ⟨c, hc, hbc⟩ := exists_lt_of_cInf_lt ht₀ (lt_div''.1 hab),
+    exact ⟨b * c, mul_mem_mul hb hc, lt_div_iff_mul_lt'.1 hbc⟩ }
+end
+
+@[to_additive] lemma cSup_div (hs₀ : s.nonempty) (hs₁ : bdd_above s) (ht₀ : t.nonempty)
+  (ht₁ : bdd_below t) :
+  Sup (s / t) = Sup s / Inf t :=
+by rw [div_eq_mul_inv, cSup_mul hs₀ hs₁ ht₀.inv ht₁.inv, cSup_inv ht₀ ht₁, div_eq_mul_inv]
+
+@[to_additive] lemma cInf_div (hs₀ : s.nonempty) (hs₁ : bdd_below s) (ht₀ : t.nonempty)
+  (ht₁ : bdd_above t) :
+  Inf (s / t) = Inf s / Sup t :=
+by rw [div_eq_mul_inv, cInf_mul hs₀ hs₁ ht₀.inv ht₁.inv, cInf_inv ht₀ ht₁, div_eq_mul_inv]
+
+end conditionally_complete_linear_order
+
+variables [complete_linear_order α] [covariant_class α α (*) (≤)]
+
+@[to_additive] lemma Sup_mul (s t : set α) : Sup (s * t) = Sup s * Sup t :=
+begin
+  refine Sup_eq_of_forall_le_of_forall_lt_exists_gt _ (λ a ha, _),
+  { rintro f ⟨a, b, ha, hb, rfl⟩,
+    exact mul_le_mul' (le_Sup ha) (le_Sup hb) },
+  { obtain ⟨b, hb, hab⟩ := lt_Sup_iff.1 (div_lt_iff_lt_mul.2 ha),
+    obtain ⟨c, hc, hbc⟩ := lt_Sup_iff.1 (div_lt''.1 hab),
+    exact ⟨b * c, mul_mem_mul hb hc, div_lt_iff_lt_mul'.1 hbc⟩ }
+end
+
+@[to_additive] lemma Inf_mul (s t : set α) : Inf (s * t) = Inf s * Inf t :=
```

Comment:
Does this hold in something weaker than a group? Is cancellative multiplication enough?

Reply:
I need to turn `Inf s * Inf t < a` into `∃ b ∈ s, b * Inf t < a`. Is cancellability enough for this?

Reply:
Actually, this seems to have nothing to do with cancellability, but rather covariance of multiplication.

Reply:
Yes it does need a group, because the key point is that division on the right is adjoint to multiplication on the left/right (see `galois_connection.l_Sup`). However there's a more general lemma to be stated in terms of `set.image2`.

Reply:
I managed to generalize (module reasonable `sorry`s) to `(conditionally_)complete_lattice` and `group`.

Reply:
And now the sorries are gone.